File: php/Operations.php
<?php
/**
 * Contains all the operations to upload, poll and download files. Unlike the Node.js scripts, this
 * operates synchronously.
 */
class Operations
{
  private $username;
  private $password;
  private $host;
  private $port;
  
  public $uploadFileName;// Set on upload
  public $pollEvery;
  public $ftp;

  /**
   * The constructor adds properties to the object which are used in init.
   *
   * @param (ftp) An instance of ftp to use.
   * @param (username) The username to get into the ftp server.
   * @param (password) The password to get into the ftp server.
   * @param (host) The host to connect to. Defaults to localhost.
   * @param (port) The port to connect to. Defaults to 21.
   * @param (polling) Number of seconds to poll for. Defaults to 300 (5 minutes) if falsey.
   */
  public function __construct(Ftp $ftp, $username, $password,
                              $host = 'localhost', $port = 21, $pollEvery = 300) 
  {
    $this->username = $username;
    $this->password = $password;
    $this->host = $host;
    $this->port = $port;
    $this->pollEvery = $pollEvery;

    if (empty($this->port)){
      $this->port = 21;
    }
    if (empty($this->host)){
      $this->host = 'localhost';
    }
    if (empty($this->pollEvery)){
      $this->pollEvery = 300;
    }

    $this->ftp = $ftp;
  }

  /**
   * Initializes the ftp object and logs in. Then goes to passive mode.
   *
   * @return TRUE if operations could be initialized.
   */
  public function init()
  {
    if (!$this->ftp->SetServer($this->host, $this->port)) {
        $this->ftp->quit();
        throw new RuntimeException("Could not set the server with $this->host:$this->port.\n");
    }

    if (!$this->ftp->connect()) {
      $this->ftp->quit();
      throw new RuntimeException("Cannot connect to $this->host:$this->port.\n");
    }

    if (!$this->ftp->login($this->username, $this->password)) {
      $this->ftp->quit();
      throw new RuntimeException(
        "Login failed with username:password $this->username:$this->password.\n"
      );
    }

    if (!$this->ftp->SetType(FTP_AUTOASCII)) {
      $this->ftp->quit();
      throw new RuntimeException("Could not set type to auto ASCII.\n");
    }

    if (!$this->ftp->Passive(TRUE)) {
      $this->ftp->quit();
      throw new RuntimeException("Could not change to passive mode.\n");
    }

    return TRUE;
  }

  /**
   * Changes to the upload directory then uploads the specified file.
   *
   * @param (file) The location of the file to upload.
   * @param (singleFile) If the file is uploaded in single file mode. Defaults to FALSE.
   *
   * @return An array of the format [<upload succeeded>, <message>].
   */
  public function upload($file, $singleFile = FALSE)
  {
    $type = $singleFile ? 'default' : 'splitfile';
    $dir = "import_{$this->username}_{$type}_config";

    $formatted = explode('/', trim($file));
    $formatted = end($formatted);

    if($this->ftp->put($formatted, "$dir/$formatted")) {
      $response_message = $this->ftp->last_message();
      if (preg_match("/.* (.*)$/", $response_message, $parsed)) {
        $this->uploadFileName = trim($parsed[1]);

        return array(TRUE, "$formatted has been uploaded as {$parsed[1]}\n");
      } else {
        return array(FALSE, "Failed to extract filename from: $response_message\n");
      }
    } else {
      $message = $this->ftp->last_message();
      return array(FALSE, "\nFile upload error: $message\n");
    }
  }

 /**
   * Polls every pollEvery seconds until the last uploaded file can be downloaded. Then downloads.
   *
   * @param (location) The absolute location to download the file to.
   * @param (self) A new version of this class to use. Defaults to $this. (For testing purposes)
   *
   * @return An array [<download succeeded>, <message>].
   */
  public function download($location, $self = '')
  {
    // So that waitAndDownload can be stubbed in the tests
    if(empty($self)){
      $self = $this;
    }

    $listing = $self->ftp->nlist('/complete');
    if (empty($listing)){
      return array(FALSE, "The /complete directory does not exist.\n");
    }

    $location = preg_replace('/\/$/', '', $location);// Remove trailing slash if present

    $formatted = $self->getDownloadFileName();
    if (array_search($formatted, $listing)){
      if(!$self->ftp->get("/complete/$formatted", "$location/$formatted")){
        $message = $self->ftp->last_message();
        return array(FALSE, "\nFile download error: $message\n");
      };

      return array(TRUE, "$formatted downloaded to $location.\n");
    } else {
      return $self->waitAndDownload($self->pollEvery, $formatted, $location);
    }
  }

  /**
   * Wait the specified time then download the file. Useful for test stubbing.
   *
   * @param (time) The time in seconds to sleep for.
   * @param (file) The filename to download. Just need the filename, no path.
   * @param (location) The absolute location to download the file to.
   * @param (self) A new version of this class to use. Defaults to $this. (For testing purposes)
   *
   * @return Result of running downloads again.
   */
  public function waitAndDownload($time, $file, $location, $self = '')
  {
    // So that echoAndSleep can be stubbed in the tests
    if(empty($self)){
      $self = $this;
    }

    $self->echoAndSleep("Waiting for results file $file ...\n", $time);

    // We could have been kicked off due to inactivity...
    $self->ftp->quit();
    if (!$self->init()){
      return array(FALSE, "Could not reconnect to the server.\n");
    }

    return $self->download($location);
  }

  // (so echo and sleep can be stubbed)
  public function echoAndSleep($message, $time)
  {
    echo($message);
    sleep($time);
  }

  /**
   * Closes the FTP connection properly. This should always be called at the end of a program using
   * this class.
   */
  public function quit()
  {
    $this->ftp->quit();
  }

  /**
   * Returns the connections details. Namely, username, password, host and port.
   *
   * @return Associative array of the username, password, host and port.
   */
  public function getConnectionDetails() 
  {
    return array(
      "username" => $this->username,
      "password" => $this->password,
      "host" => $this->host,
      "port" => $this->port,
    );
  }

  /**
   * Retrieves the upload file name and transforms it to the download one. 
   *
   * @return The current download name of the current upload.
   */
  public function getDownloadFileName()
  {
    if (empty($this->uploadFileName)){
      return $this->uploadFileName;
    }

    $formatted = preg_replace('/source_/', 'archive_', $this->uploadFileName, 1);

    if (strpos($formatted, '.csv') || strpos($formatted, '.txt')){
      $formatted = substr($formatted, 0, -4) .'.zip';
    }

    return $formatted;
  }
}
?>

File: dotnet/CSharpFTPExample/Operations.cs
﻿using System;
using System.Linq;
using System.Collections.Specialized;
using System.Net;
using System.IO;
using System.Reflection;

namespace CSharpFTPExample
{
    /// <summary>
    /// Used by the Program.cs file to load and download files. This can be directly required to integrate into
    /// your own .Net application.
    /// </summary>
    public class Operations
    {
        private string username;
        private string password;
        private string host;
        private int port;

        public int pollEvery;
        public IWebClient ftp = null;

        /// <summary>
        /// The constructor adds properties to the object which are used in init.
        /// <param name="username">The username to get into the ftp server.</param>
        /// <param name="password">The password to get into the ftp server.</param>
        /// <param name="host">The host to connect to.</param>
        /// <param name="host">The port to connect to.</param>
        /// <param name="pollEvery">Number of seconds to poll for.</param>
        /// </summary>
        public Operations(string username, string password, string host = "localhost", int port = 21, int pollEvery = 300)
        {
            this.username = username;
            this.password = password;
            this.host = host;
            this.port = port;
            this.pollEvery = pollEvery;
        }

        /// <summary>
        /// Initializes the web client was initialized with the correct credentials. Returns
        /// <value>true if this succeeded.</value>
        /// </summary>
        public Boolean Init()
        {
            this.ftp = new WrappedWebClient();

            ftp.Credentials = new NetworkCredential(username, password);

            return true;
        }

        /// <summary>
        /// Uploads the given filename to the FTP server using the information sent into initialization.
        /// <param name="filename">The absolute path name of the file.</param>
        /// <value>true if this succeeded.</value>
        /// </summary>
        public Boolean Upload(string filename)
        {
            var type = "splitfile";
            var directory = "/import_" + username + "_" + type + "_config/";

            try
            {
                ftp.UploadFile("ftp://" + host + ':' + port + directory + new FileInfo(filename).Name, "STOR", filename);

                //string[] status = GetStatusDescription(ftp);
                //Console.WriteLine("Description: " + status[0] + " " + status[1]);
            }
            catch (Exception e)
            {
                throw new Exception(e + "");
            }

            return true;
        }

        /// <summary>
        /// Returns
        /// <value>an ordered dictionary of the username, password, host and port.</value>
        /// </summary>
        public OrderedDictionary GetConnectionDetails()
        {
            OrderedDictionary dictionary = new OrderedDictionary();
            dictionary.Add("username", username);
            dictionary.Add("password", password);
            dictionary.Add("host", host);
            dictionary.Add("port", port);

            return dictionary;
        }

        /// <summary>
        /// A way of extracting ftp responses from WebClient obtained from http://stackoverflow.com/a/6470446. Returns
        /// <value>An array in the form [<status code>, <description>]</value>
        /// </summary>
        private string[] GetStatusDescription(IWebClient client)
        {
            FieldInfo responseField = client.GetType().BaseType.GetField("m_WebResponse", BindingFlags.Instance | BindingFlags.NonPublic);

            if (responseField != null)
            {
                FtpWebResponse response = responseField.GetValue(client) as FtpWebResponse;

                if (response != null)
                {
                    return new string[] { (int)response.StatusCode + "", response.StatusDescription };
                }
                else
                {
                    throw new Exception("Error: Could not get the response from the server.");
                }
            }

            throw new Exception("Error: Could not get the response field from the server.");
        }
    }
}
